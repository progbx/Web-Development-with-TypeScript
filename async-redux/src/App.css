.App {
    text-align: center;
    font-family: Arial, sans-serif;
    margin: 20px;
    max-width: 600px;
    margin: auto;
}

h1 {
    color: #333;
}

.form {
    margin-bottom: 20px;
}

input {
    margin: 5px;
    padding: 10px;
    border: 1px solid #ccc;
    border-radius: 4px;
    font-size: 16px;
    width: calc(50% - 22px); /* Adjust width to fit within the form */
}

input[type="text"] {
    width: calc(50% - 22px); /* Adjust width to fit within the form */
}

button {
    padding: 10px 20px;
    border: none;
    border-radius: 4px;
    background-color: #007bff;
    color: white;
    cursor: pointer;
    font-size: 16px;
}

button:hover {
    background-color: #0056b3;
}

button:active {
    filter: opacity(90%);
}

.contact-list {
    list-style: none;
    padding: 0;
}

.contact-list li {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin: 10px 0;
    padding: 10px;
    border: 1px solid #ccc;
    border-radius: 4px;
    background-color: #f9f9f9;
}

.contact-list span {
    font-size: 16px;
}

.contact-list button {
    background-color: #dc3545;
    font-size: 14px;
}

.contact-list button:hover {
    background-color: #c82333;
}

p {
    color: #dc3545; /* Error color */
}

.add-contact-button {
    margin-top: 1rem;
}

.info {
    width: 100%;
    margin-bottom: 1rem;
}
.info button {
    margin-left: 1rem;
}

.info-last-action {
    padding: 10px;
    border: 1px solid #ccc;
    border-radius: 4px;
    background-color: #f9f9f9;
}