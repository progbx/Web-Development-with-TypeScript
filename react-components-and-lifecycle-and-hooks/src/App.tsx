import ValidatedInput from "./components/ValidatedInput";
import TodoList from "./components/TodoList.tsx";

function App() {
  const validateLength = (value: string) => value.length >= 5;

  return (
    <>
        <h1>Validated Input</h1>
        <ValidatedInput
            validationFunction={validateLength}
            errorMessage="Input is too short (minimum 5 characters)"
        />
        <TodoList/>
    </>
  );
}

export default App;
