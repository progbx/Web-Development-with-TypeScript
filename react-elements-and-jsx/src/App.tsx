import Title from './components/Title';
import PlanetsList from './components/PlanetsList';

function App() {
  return (
    <div>
      <Title />
      <PlanetsList />
    </div>
  );
}

export default App;
