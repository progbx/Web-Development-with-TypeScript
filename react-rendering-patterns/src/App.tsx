
import { currentWeather, forecastData } from "./data/weatherData";
import WeatherCard from "./components/WeatherCard";
import Forecast from "./components/Forecast";

const App = () => {
    return (
        <section>
            <h1>Let's talk weather!</h1>

            <WeatherCard weather={currentWeather} />
            <Forecast forecast={forecastData} />
        </section>
    );
};

export default App;
