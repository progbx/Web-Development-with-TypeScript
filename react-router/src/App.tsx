import { BrowserRouter as Router, Routes, Route, Link } from 'react-router-dom';
import Home from './components/Home';
import About from './components/About';
import Gallery from './components/Gallery';
import PhotoCard from './components/PhotoCard';
import NotFound from './components/NotFound';

function App() {
  return (
    <Router>
      <nav>
        <Link to="/">Home</Link>
        <Link to="/about">About</Link>
        <Link to="/gallery">Gallery</Link>
      </nav>
      <Routes>
        <Route path="/" element={<Home />} />
        <Route path="/about" element={<About />} />
        <Route path="/gallery" element={<Gallery />} />
        <Route path="/gallery/:id" element={<PhotoCard />} />
        <Route path="*" element={<NotFound />} />
      </Routes>
    </Router>
  );
}

export default App;
