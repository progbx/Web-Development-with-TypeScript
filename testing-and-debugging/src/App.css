body {
    font-family: monospace sans-serif;
}


.App {
    text-align: center;
}

.App-header {
    background-color: #282c34;
    min-height: 100vh;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: flex-start;
    font-size: calc(10px + 2vmin);
    color: white;
    padding: 0 20px;
    position: relative; /* Ensure positioning context for the menu container */
}

.App-header h1 {
    font-size: 2.5em;
    margin: 0;
}

.App-header p {
    font-size: 1.2em;
    margin-top: 10px;
    color: #b0b0b0;
}

button {
    margin-top: 20px;
    padding: 10px 20px;
    font-size: 16px;
    cursor: pointer;
    background-color: #61dafb;
    border: none;
    border-radius: 5px;
    color: white;
}

button:hover {
    background-color: #4fa3d1;
}

.App-menu-container {
    position: fixed; /* Fixes the menu to a specific position on the viewport */
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background-color: rgba(0, 0, 0, 0.8); /* Semi-transparent background */
    display: flex;
    justify-content: center;
    align-items: center;
    z-index: 1000; /* Ensures the menu appears above other content */
}

.App-menu {
    background-color: #282c34;
    padding: 20px;
    border-radius: 8px;
}

.App-menu ul {
    list-style-type: none;
    padding: 0;
    margin: 0;
}

.App-menu li {
    margin: 15px 0;
}

.App-menu a {
    color: #61dafb;
    text-decoration: none;
    font-weight: bold;
    font-size: 1.2em;
}

.App-menu a:hover {
    text-decoration: underline;
}
