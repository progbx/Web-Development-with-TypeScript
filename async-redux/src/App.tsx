
import React, { useState, ChangeEvent } from "react";
import "./App.css";
import { useGetContactsQuery, useAddContactMutation, useRemoveContactMutation } from "./features/contactsApi";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "./redux/store";
import { setLastAction, resetLastAction } from "./features/infoSlice";

const App: React.FC = () => {
    const [name, setName] = useState<string>("");
    const [phone, setPhone] = useState<string>("");
    const dispatch = useDispatch();

    const { data: contacts, isLoading, refetch } = useGetContactsQuery();
    const contactsList = Array.isArray(contacts) ? contacts : [];
    const [addContact] = useAddContactMutation();
    const [removeContact] = useRemoveContactMutation();
    const lastAction = useSelector((state: RootState) => state.info.lastAction);

    const processResetsetLastAction = () => {
        dispatch(resetLastAction());
    };

    const handleAddContact = async () => {
        if (name && phone) {
            try {
                await addContact({ name, phone }).unwrap();
                setName("");
                setPhone("");
                await refetch();
                dispatch(setLastAction("Add Contact"));
            } catch (err) {
                console.error("Failed to add contact:", err);
            }
        }
    };

    const handleRemoveContact = async (id: string) => {
        try {
            await removeContact(id).unwrap();
            await refetch();
            dispatch(setLastAction("Remove Contact"));
        } catch (err) {
            console.error("Failed to remove contact:", err);
        }
    };

    return (
        <div className="App">
            <h1>Contact List </h1>
            <div className="info">
                <span className="info-last-action">
                    {lastAction === null
                        ? "No Updates in This Session"
                        : `Last Action: ${lastAction}`}
                </span>
                <button onClick={processResetsetLastAction}>Reset</button>
            </div>
            <div className="form">
                <div>
                    <label>
                        Contact Name:
                        <input
                            name="contact-name"
                            type="text"
                            value={name}
                            onChange={(e: ChangeEvent<HTMLInputElement>) => setName(e.target.value)}
                            placeholder="Contact Name"
                        />
                    </label>
                </div>

                <div>
                    <label>
                        Phone Number:
                        <input
                            name="contact-phone"
                            type="text"
                            value={phone}
                            onChange={(e: ChangeEvent<HTMLInputElement>) => setPhone(e.target.value)}
                            placeholder="Phone Number"
                        />
                    </label>
                </div>

                <button
                    className="add-contact-button"
                    onClick={handleAddContact}
                >
                    Add Contact
                </button>
            </div>
            {isLoading ? (
                <p>Loading...</p>
            ) : (
                <ul className="contact-list">
                    {contactsList.map((contact) => (
                        <li key={contact.id}>
                            <span>
                                {contact.name} - {contact.phone}
                            </span>
                            <button
                                onClick={() => handleRemoveContact(contact.id)}
                            >
                                Remove
                            </button>
                        </li>
                    ))}
                </ul>
            )}
        </div>
    );
};

export default App;
