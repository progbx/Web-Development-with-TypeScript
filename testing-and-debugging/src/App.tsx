import React, { useState } from 'react';
import './App.css';

const App: React.FC = () => {
  const [isMenuOpen, setIsMenuOpen] = useState<boolean>(false);

  const toggleMenu = () => {
    setIsMenuOpen(prevState => !prevState);
  };

  return (
    <div className="App">
      <header className="App-header">
        <h1>Empower Your Journey with Our Solutions</h1>
        <p>Discover tools and services designed to streamline your life.</p>
        <button onClick={toggleMenu}>
          {isMenuOpen ? 'Close Menu' : 'Open Menu'}
        </button>
        {isMenuOpen && (
          <nav className="App-menu">
            <ul>
              <li><a href="#home">Home</a></li>
              <li><a href="#about">About Us</a></li>
              <li><a href="#services">Our Services</a></li>
              <li><a href="#contact">Contact Us</a></li>
            </ul>
          </nav>
        )}
      </header>
    </div>
  );
};

export default App;
