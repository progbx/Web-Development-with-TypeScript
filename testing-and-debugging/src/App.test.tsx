import "@testing-library/jest-dom";
import { render, screen, fireEvent, act } from "@testing-library/react";
import App from "./App";

describe("App", () => {
    it("should match snapshot", () => {
        const { asFragment } = render(<App />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("should match snapshot after openning a menu", () => {
        const { asFragment, getByRole } = render(<App />);
        fireEvent.click(getByRole("button"));
        expect(asFragment()).toMatchSnapshot();
    });

    it("should have a correct page heading", () => {
        render(<App />);
        const heading = screen.getByRole("heading", { name: /Empower Your Journey with Our Solutions/i });
        expect(heading).toBeInTheDocument();
    });

    it("should change button text after clicking", () => {
        render(<App />);
        const button = screen.getByRole("button");
        fireEvent.click(button);
        expect(button).toHaveTextContent("Close Menu");
    });

    it("should change button text after clicking twice", () => {
        render(<App />);
        const button = screen.getByRole("button");
        fireEvent.click(button);
        fireEvent.click(button);
        expect(button).toHaveTextContent("Open Menu");
    });

    it("should have a menu after clicking the button", () => {
        render(<App />);
        const button = screen.getByRole("button");
        fireEvent.click(button);
        const nav = screen.getByRole("navigation");
        expect(nav).toBeInTheDocument();
    });

    it("should have four menu items", () => {
        render(<App />);
        const button = screen.getByRole("button");
        fireEvent.click(button);
        const nav = screen.getByRole("navigation");
        const list = nav.querySelector("ul");
        expect(list).toBeInTheDocument();
        const items = list?.querySelectorAll("li");
        expect(items).toHaveLength(4);
    });
});
