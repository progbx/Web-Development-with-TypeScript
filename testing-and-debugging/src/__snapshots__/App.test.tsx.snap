// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <h1>
        Empower Your Journey with Our Solutions
      </h1>
      <p>
        Discover tools and services designed to streamline your life.
      </p>
      <button>
        Open Menu
      </button>
    </header>
  </div>
</DocumentFragment>
`;

exports[`App should match snapshot after openning a menu 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <h1>
        Empower Your Journey with Our Solutions
      </h1>
      <p>
        Discover tools and services designed to streamline your life.
      </p>
      <button>
        Close Menu
      </button>
      <nav
        class="App-menu"
      >
        <ul>
          <li>
            <a
              href="#home"
            >
              Home
            </a>
          </li>
          <li>
            <a
              href="#about"
            >
              About Us
            </a>
          </li>
          <li>
            <a
              href="#services"
            >
              Our Services
            </a>
          </li>
          <li>
            <a
              href="#contact"
            >
              Contact Us
            </a>
          </li>
        </ul>
      </nav>
    </header>
  </div>
</DocumentFragment>
`;
